From aa7137c1ef26781e048b6d6c60d9c8d16336b690 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Fri, 8 Mar 2024 22:29:07 +0100
Subject: [PATCH] sushi-media-bin: Fix llvmpipe detection

Recent versions of mesa have removed the "Gallium on" prefix from the renderer string. Also detect "softpipe" as a software GL rasterizer, and introduce a new environment variable "SUSHI_USE_GST_GTKSINK" to force gtksink.
---
 src/libsushi/sushi-media-bin.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/src/libsushi/sushi-media-bin.c b/src/libsushi/sushi-media-bin.c
index 585db5d..5f8cbc2 100644
--- a/src/libsushi/sushi-media-bin.c
+++ b/src/libsushi/sushi-media-bin.c
@@ -510,6 +510,9 @@ sushi_media_bin_gl_check (GtkWidget *widget)
 {
   static gsize gl_works = 0;
 
+  if (g_getenv ("SUSHI_USE_GST_GTKSINK") != NULL)
+    return FALSE;
+
   if (g_once_init_enter (&gl_works))
     {
       GError *error = NULL;
@@ -533,8 +536,8 @@ sushi_media_bin_gl_check (GtkWidget *widget)
             GST_WARNING ("nouveau is blacklisted, since sharing gl contexts in "
                          "multiple threads is not supported "
                          "and will eventually make it crash.");
-          else if (g_strstr_len (renderer, -1, "Gallium") &&
-                   g_strstr_len (renderer, -1, "llvmpipe"))
+          else if (g_strstr_len (renderer, -1, "llvmpipe") ||
+                   g_strstr_len (renderer, -1, "softpipe"))
             GST_INFO ("Detected software GL rasterizer, falling back to gtksink");
           else
             works = 2;
-- 
GitLab

